Review

BlackBeltPanda opened the pull request:
PR for code review

BlackBeltPanda committed:
Update project names

BlackBeltPanda committed:
Add components and functionality

BlackBeltPanda committed:
Refactor

Applying what I've learned to everything

BlackBeltPanda committed:
Add Size and Smoothing Sliders and Clear Button

Create Slider component
Interpolate in real-time instead of on mouse release
Refactor to simplify behavior between whiteboard and canvas
Enable Kotlin explicit API and explicitly declare visibility and types
Reformat according to ktfmt

Sk1er commented on src/main/kotlin/gg/essential/elementaPractice/gui/WhiteboardGUI.kt:
```diff
@@ -0,0 +1,205 @@
+        // Save/Load buttons?
+
+        // Button to close the whiteboard
+        val closeButton = UIBlock(Color(207, 207, 196)).constrain {
```
When working on Essential UIs, we use a class to store all the colors for UIs so that if we want to change a color, we only need to change it in one place instead of every place it's used. Check out the VigilancePalette and EssentialPalette classes.

Sk1er commented on src/main/kotlin/gg/essential/elementaPractice/gui/WhiteboardGUI.kt:
```diff
@@ -0,0 +1,205 @@
+        UIText("X", shadow = false).constrain {
+            x = CenterConstraint()
+            y = CenterConstraint()
+            textScale = 0.5.pixels()
```
We try to avoid fractional scales for text because the scaling can be really weird and introduce artifacts. The Minecraft font renderer isn't build to handle scaling for non integets

Sk1er commented on src/main/kotlin/gg/essential/elementaPractice/gui/WhiteboardGUI.kt:
```diff
@@ -0,0 +1,205 @@
+        // Size slider label
+        UIText("Marker Size", shadow = false).constrain {
+            x = CenterConstraint()
+            y = (-5).pixels()
```
The children of a component should render inside of it

Sk1er commented on src/main/kotlin/gg/essential/elementaPractice/gui/WhiteboardGUI.kt:
```diff
@@ -0,0 +1,205 @@
+        for ((index, colorTool) in enumValues<ColorTools>().withIndex()) {
+            val tool = ColorToolComponent(colorTool).constrain {
+                // Shift by 34 pixels (tool width + 4) for each tool and add 4 pixel padding
+                x = (index * 34 + 4).pixels()
```
This would be a good place to use a SiblingConstraint()

Sk1er commented on src/main/kotlin/gg/essential/elementaPractice/gui/components/ColorToolComponent.kt:
```diff
@@ -0,0 +1,59 @@
+        }
+
+        onMouseClick {
+            setWidth(32.pixels())
```
This is likely redundant because in order for the component to be clicked, the hover event will be fired and the width will already be set to 32

Sk1er commented on src/main/kotlin/gg/essential/elementaPractice/gui/components/SliderComponent.kt:
```diff
@@ -0,0 +1,88 @@
+
+        // Slider bar/knob
+        val slider = UIBlock(Color(150, 150, 150).toConstraint()).constrain {
+            x = (getSegmentPosition(currentValue - oneBased.toInt()) - 1.5).pixels()
+            width = 3.pixels()
+            height = 8.pixels()
+        } childOf this
+
+        // Value indicator
+        val indicator = UIText(currentValue.toString()).constrain {
+            x = 23.pixels()
+            y = CenterConstraint()
+            color = Color.WHITE.toConstraint()
+            textScale = 0.5.pixels()
+        } childOf bar
+
+        onMouseDrag { mouseX, mouseY, mouseButton ->
+            if (mouseButton != 0) return@onMouseDrag
+            if (mouseX !in 0f..bar.getWidth() || mouseY !in 0f..slider.getHeight()) return@onMouseDrag
+
+            // Calculate the width of each segment between values
+            val segWidth = bar.getWidth() / (numValues - 1)
+            // Round the x position to the nearest segment edge
+            val segRounded = (mouseX / segWidth).roundToInt() * segWidth
+
+            slider.setX((segRounded - 1.5).pixels())
+
+            // Set the current value to the index of the segment edge
+            currentValue = (segRounded / segWidth).roundToInt() + oneBased.toInt()
+
+            // Update indicator text
+            indicator.setText(currentValue.toString())
+        }
```
This would be a good place to use a State object. 

The initial state would be set to the currentValue constructor parameter and that currentValue would be made immutable. 

In on mouse drag, you could then update the value of the state to what it should be. 

You can call 
```kotlin
state.onSetValueAndNow {
    slider.setX(//impl)
    indicator.setText(it.toString())
}
```

BlackBeltPanda committed:
Review changes

Utilize palette for colors and icons
Move slider label to slider component
Better component constraints
Check tick phase during tick event
Don't use fractional text scaling

BlackBeltPanda committed:
Utilize State Object for Slider Value

Allows line width and smoothing values to update when the state updates

BlackBeltPanda committed:
Add Button Component

Replace close and clear buttons with button components

BlackBeltPanda committed:
Add ability to draw points on single click

asbyth commented on src/main/kotlin/gg/essential/elementaPractice/ElementaPractice.kt:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    // Register whiteboard command
+    @Mod.EventHandler
+    public fun init(event: FMLServerStartingEvent) {
+        event.registerServerCommand(WhiteboardCommand())
+    }
```
this should go in the `FMLInitializationEvent` event method as `ClientCommandHandler.instance.registerCommand(class)` instead
